### PR TITLE
Vastly improve the printer

### PR DESCRIPTION
**Pull request by AltGr**
* fix lots of Format boxes
* add parens based on precedence in the printer
* fix interaction of Format and the colored line debug tags

### CODE REVIEW

File: compiler/shared_ast/print.ml [hunk 1]
```diff
@@ -267,8 +273,105 @@ let var_debug fmt v =
 
 let var fmt v = Format.pp_print_string fmt (Bindlib.name_of v)
 
-let needs_parens (type a) (e : (a, _) gexpr) : bool =
-  match Marked.unmark e with EAbs _ | EStruct _ -> true | _ -> false
+(* Define precedence levels for auto parentheses *)
+module Precedence = struct
+  type op = Xor | And | Or | Comp | Mul | Div | Add | Sub
```
Comment (denismerigoux) [inline compiler/shared_ast/print.ml:278]: Yeah parenthesis precedence !!!

File: compiler/shared_ast/print.ml [hunk 2]
```diff
@@ -282,29 +385,38 @@ let rec expr_aux :
...
   let var = if debug then var_debug else var in
-  let with_parens fmt e =
-    if needs_parens e then (
+  let rec skip_log : type a. (a, 't) gexpr -> (a, 't) gexpr = function
+    | EApp { f = EOp { op = Log _; _ }, _; args = [e] }, _ -> skip_log e
+    | e -> e
+  in
+  let e = if not debug then skip_log e else e in
+  let paren ~rhs expr fmt e1 =
+    if Precedence.needs_parens ~rhs ~context:e (skip_log e1) then (
```
Comment (denismerigoux) [inline compiler/shared_ast/print.ml:394]: We want to skip logs only when we're not in debug mode (`-d`) no?

File: tests/test_array/good/aggregation_2.catala_en [hunk 3]
```diff
@@ -33,13 +33,14 @@ scope B:
...
+             {S "id" = 1; "income" = $9.00};
+             {S "id" = 2; "income" = $5.20} ]
 ```
 
 ```catala-test-inline
 $ catala Interpret -s B
 [RESULT] Computation successful! Results:
-[RESULT] argmax = S { "id"= 1; "income"= $9.00 }
-[RESULT] argmin = S { "id"= 0; "income"= $0.00 }
+[RESULT] argmax = {S "id" = 1; "income" = $9.00}
```
Comment (denismerigoux) [inline tests/test_array/good/aggregation_2.catala_en:44]: This printing behavior is a bit weird. The previous one was influenced by Rust's `struct`. Maybe `S.{ ... }` ?

**Review by denismerigoux — APPROVED** [review]
Thanks so much @AltGr ! This was long overdue.

File: tests/test_array/good/aggregation_2.catala_en (hunk 3 above)
Comment (AltGr) [reply tests/test_array/good/aggregation_2.catala_en:44]: Yeah this was inspired by OCaml's `{ Foo.bar = _; baz = _ }` but in that case `Foo` is a module; I have no strong opinion on this, whatever is more readable and understandable.

Comment (denismerigoux) [reply tests/test_array/good/aggregation_2.catala_en:44]: @adelaett has another take in #412, `{ ...}_S`

**Pull request merged**